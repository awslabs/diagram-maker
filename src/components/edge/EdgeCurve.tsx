import * as classnames from 'classnames';
import * as Preact from 'preact';

import { getInflectionPoint } from 'diagramMaker/service/positionUtils';
import { Position } from 'diagramMaker/state/types';

// @NOTE: Depending on node connector placement
// the bezier curve is drawn differently, so we have two types
export enum EdgeStyle {
  LEFT_RIGHT_BEZIER,
  TOP_BOTTOM_BEZIER,
  STRAIGHT,
  QUADRATIC_BEZIER
}

export interface EdgeCurveProps {
  src: Position;
  dest: Position;
  edgeStyle: EdgeStyle;
  className?: string;
  showArrowhead?: boolean;
}

const getQuadraticBezierPath = (src: Position, dest: Position, className?: string, markerEnd?: string): JSX.Element => {
  const { x: x1, y: y1 } = src;
  const { x: x2, y: y2 } = dest;
  const { x: startingX, y: startingY } = getInflectionPoint(src, dest);
  const classNames = classnames('dm-path', className);

  const pathData = `M${x1},${y1} Q${startingX},${startingY} ${x2},${y2}`;
  return <path className={classNames} d={pathData} marker-end={markerEnd} />;
};

const getLeftRightBezierPath = (src: Position, dest: Position, className?: string, markerEnd?: string): JSX.Element => {
  const { x: x1, y: y1 } = src;
  const { x: x2, y: y2 } = dest;
  const width = x2 - x1;
  const halfWidth = width / 2;
  const classNames = classnames('dm-path', className);

  const pathData = `M${x1},${y1} C${x1 + Math.abs(halfWidth)},${y1} ${x2 - Math.abs(halfWidth)},${y2} ${x2},${y2}`;
  return <path className={classNames} d={pathData} marker-end={markerEnd} />;
};

const getTopBottomBezierPath = (src: Position, dest: Position, className?: string, markerEnd?: string): JSX.Element => {
  const { x: x1, y: y1 } = src;
  const { x: x2, y: y2 } = dest;
  const height = y2 - y1;
  const halfHeight = height / 2;
  const classNames = classnames('dm-path', className);

  const pathData = `M${x1},${y1} C${x1},${y1 + Math.abs(halfHeight)} ${x2},${y2 - Math.abs(halfHeight)} ${x2},${y2}`;
  return <path className={classNames} d={pathData} marker-end={markerEnd} />;
};

const getStraightPath = (src: Position, dest: Position, className?: string, markerEnd?: string): JSX.Element => {
  const classNames = classnames('dm-path', className);
  return <line className={classNames} x1={src.x} y1={src.y} x2={dest.x} y2={dest.y} marker-end={markerEnd} />;
};

const EdgeCurve = (props: Preact.RenderableProps<EdgeCurveProps>): JSX.Element => {
  const { src, dest, edgeStyle, className, showArrowhead } = props;
  let markerEnd;
  if (showArrowhead) {
    markerEnd = 'url(#arrow)';
  }
  switch (edgeStyle) {
    case EdgeStyle.LEFT_RIGHT_BEZIER:
      return getLeftRightBezierPath(src, dest, className, markerEnd);
    case EdgeStyle.TOP_BOTTOM_BEZIER:
      return getTopBottomBezierPath(src, dest, className, markerEnd);
    case EdgeStyle.STRAIGHT:
      return getStraightPath(src, dest, className, markerEnd);
    case EdgeStyle.QUADRATIC_BEZIER:
      return getQuadraticBezierPath(src, dest, className, markerEnd);
  }
};

export default EdgeCurve;
