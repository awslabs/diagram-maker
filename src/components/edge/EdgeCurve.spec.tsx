import * as Preact from 'preact';
import { shallow } from 'preact-render-spy';

import { default as EdgeCurve, EdgeStyle } from './EdgeCurve';

describe('getCurve', () => {

  describe('left/right placement', () => {
    it('renders at given points with x1 < x2, y1 > y2', () => {
      const src = { x:0, y:100 };
      const dest = { x:100, y:0 };
      const edgeStyle = EdgeStyle.LEFT_RIGHT_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with x1 < x2, y1 < y2', () => {
      const src = { x:0, y:0 };
      const dest = { x:100, y:100 };
      const edgeStyle = EdgeStyle.LEFT_RIGHT_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with y1 = y2', () => {
      const src = { x:0, y:0 };
      const dest = { x:100, y:0 };
      const edgeStyle = EdgeStyle.LEFT_RIGHT_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with x1 = x2', () => {
      const src = { x:0, y:0 };
      const dest = { x:0, y:100 };
      const edgeStyle = EdgeStyle.LEFT_RIGHT_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });
  });

  describe('top/bottom placement', () => {
    it('renders at given points with x1 < x2, y1 > y2', () => {
      const src = { x:0, y:100 };
      const dest = { x:100, y:0 };
      const edgeStyle = EdgeStyle.TOP_BOTTOM_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with x1 < x2, y1 < y2', () => {
      const src = { x:0, y:0 };
      const dest = { x:100, y:100 };
      const edgeStyle = EdgeStyle.TOP_BOTTOM_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with y1 = y2', () => {
      const src = { x:0, y:0 };
      const dest = { x:100, y:0 };
      const edgeStyle = EdgeStyle.TOP_BOTTOM_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with x1 = x2', () => {
      const src = { x:0, y:0 };
      const dest = { x:0, y:100 };
      const edgeStyle = EdgeStyle.TOP_BOTTOM_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });
  });

  describe('straight curve', () => {
    it('renders at given points', () => {
      const src = { x:0, y:100 };
      const dest = { x:100, y:0 };
      const edgeStyle = EdgeStyle.STRAIGHT;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });
  });

  describe('arrowhead marker', () => {
    it('renders when show arrowhead is passed', () => {
      const src = { x:0, y:100 };
      const dest = { x:100, y:0 };
      const edgeStyle = EdgeStyle.STRAIGHT;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
          showArrowhead={true}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });
  });

  describe('quadratic bezier', () => {
    it('renders at given points with x1 < x2, y1 > y2', () => {
      const src = { x:0, y:100 };
      const dest = { x:100, y:0 };
      const edgeStyle = EdgeStyle.QUADRATIC_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with x1 < x2, y1 < y2', () => {
      const src = { x:0, y:0 };
      const dest = { x:100, y:100 };
      const edgeStyle = EdgeStyle.QUADRATIC_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with y1 = y2', () => {
      const src = { x:0, y:0 };
      const dest = { x:100, y:0 };
      const edgeStyle = EdgeStyle.QUADRATIC_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });

    it('renders at given points with x1 = x2', () => {
      const src = { x:0, y:0 };
      const dest = { x:0, y:100 };
      const edgeStyle = EdgeStyle.QUADRATIC_BEZIER;
      const edgeCurve = shallow(
        <EdgeCurve
          src={src}
          dest={dest}
          edgeStyle={edgeStyle}
        />
      );

      expect(edgeCurve).toMatchSnapshot();
    });
  });

});
