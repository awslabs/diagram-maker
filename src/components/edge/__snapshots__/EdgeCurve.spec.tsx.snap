// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCurve arrowhead marker renders when show arrowhead is passed 1`] = `
preact-render-spy (1 nodes)
-------
<line
  class="dm-path"
  x1={0}
  y1={100}
  x2={100}
  y2={0}
  marker-end="url(#arrow)"
>
</line>

`;

exports[`getCurve left/right placement renders at given points with x1 < x2, y1 < y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 C50,0 50,100 100,100"
>
</path>

`;

exports[`getCurve left/right placement renders at given points with x1 < x2, y1 > y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,100 C50,100 50,0 100,0"
>
</path>

`;

exports[`getCurve left/right placement renders at given points with x1 = x2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 C0,0 0,100 0,100"
>
</path>

`;

exports[`getCurve left/right placement renders at given points with y1 = y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 C50,0 50,0 100,0"
>
</path>

`;

exports[`getCurve quadratic bezier renders at given points with x1 < x2, y1 < y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 Q100,0 100,100"
>
</path>

`;

exports[`getCurve quadratic bezier renders at given points with x1 < x2, y1 > y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,100 Q0,0 100,0"
>
</path>

`;

exports[`getCurve quadratic bezier renders at given points with x1 = x2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 Q50,50 0,100"
>
</path>

`;

exports[`getCurve quadratic bezier renders at given points with y1 = y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 Q50,-50 100,0"
>
</path>

`;

exports[`getCurve straight curve renders at given points 1`] = `
preact-render-spy (1 nodes)
-------
<line
  class="dm-path"
  x1={0}
  y1={100}
  x2={100}
  y2={0}
>
</line>

`;

exports[`getCurve top/bottom placement renders at given points with x1 < x2, y1 < y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 C0,50 100,50 100,100"
>
</path>

`;

exports[`getCurve top/bottom placement renders at given points with x1 < x2, y1 > y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,100 C0,150 100,-50 100,0"
>
</path>

`;

exports[`getCurve top/bottom placement renders at given points with x1 = x2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 C0,50 0,50 0,100"
>
</path>

`;

exports[`getCurve top/bottom placement renders at given points with y1 = y2 1`] = `
preact-render-spy (1 nodes)
-------
<path
  class="dm-path"
  d="M0,0 C0,0 100,0 100,0"
>
</path>

`;
